src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.MetadataDb.cs

```diff
@@ -120,30 +125,17 @@ internal MetadataDb(int payloadLength)
                     initialSize = OneMegabyte;
                 }
 
-                _data = ArrayPool<byte>.Shared.Rent(initialSize);
-                Length = 0;
-#if DEBUG
-                _isLocked = false;
-#endif
+                byte[] data = ArrayPool<byte>.Shared.Rent(initialSize);
+                return new MetadataDb(data, isLocked: false, convertToAlloc);
             }
 
-            internal MetadataDb(MetadataDb source, bool useArrayPools)
+            internal static MetadataDb CreateLocked(int payloadLength)
             {
-                Length = source.Length;
-
-#if DEBUG
-                _isLocked = !useArrayPools;
-#endif
+                // Add one row worth of data since we need at least one row for a primitive type.
+                int size = payloadLength + DbRow.Size;
 
-                if (useArrayPools)
-                {
-                    _data = ArrayPool<byte>.Shared.Rent(Length);
-                    source._data.AsSpan(0, Length).CopyTo(_data);
-                }
-                else
-                {
-                    _data = source._data.AsSpan(0, Length).ToArray();
-                }
+                byte[] data = new byte[size];
+                return new MetadataDb(data, isLocked: true, convertToAlloc: false);
```

Comment:
I am a bit confused about the relationship between create rented and createlocked and convert to alloc. Why does `CreateLocked` mean `convertToAlloc` is false? It looks like, in most cases, `CreateRented` would also have `convertToAlloc` be false.

Reply:
Scenarios:
1) `isLocked`=true + `convertToAlloc`=false. When called by pre-existing "Clone()" path (the buffer is known already).
2) `isLocked`=false+ `convertToAlloc`=false. Basically the pre-existing path with the exception of the "Clone()" path.
3) `isLocked`=true + `convertToAlloc`=false. New fast path for primitives (not arrays and objects). The size is known from the provided JSON length and the provided buffer is not pooled.
4) `isLocked`=false+ `convertToAlloc`=true. New fast path for arrays and objects. The size is not yet known, so we want to use a pooled buffer until we know the final size, and then release the pooled buffer after creating a non-pooled alloc.

Reply:
Thanks for the clarification. I will come back and re-read the combination to understand it better.

Maybe parts of this explanation should be added to comments in source.

Reply:
I added a table in comments to doc the behavior